dot/sync/chain_processor_test.go

```diff
@@ -0,0 +1,607 @@
+
+package sync
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/runtime/mocks"
+	"github.com/ChainSafe/gossamer/lib/runtime/storage"
+	"github.com/ChainSafe/gossamer/lib/trie"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_chainProcessor_handleBlock_baseCase(t *testing.T) {
+	t.Parallel()
+
+	ctrl := gomock.NewController(t)
+
+	block := &types.Block{
+		Header: types.Header{
+			Number: 0,
+		},
+		Body: types.Body{},
+	}
+
+	mockBlockState := NewMockBlockState(ctrl)
+	mockHeader := &types.Header{
+		Number:    0,
+		StateRoot: trie.EmptyHash,
+	}
+	mockHeaderHash := mockHeader.Hash()
+	mockBlock := &types.Block{
+		Header: types.Header{
+			Number: 0,
+		},
+		Body: types.Body{},
+	}
+	mockTrieState, err := storage.NewTrieState(nil)
+	require.NoError(t, err)
+	mockBlockState.EXPECT().GetHeader(common.Hash{}).Return(mockHeader, nil)
+	mockInstance := mocks.NewMockInstance(ctrl)
+	mockInstance.EXPECT().SetContextStorage(mockTrieState)
+	mockInstance.EXPECT().ExecuteBlock(mockBlock)
+	mockBlockState.EXPECT().GetRuntime(&mockHeaderHash).Return(mockInstance, nil)
+
+	mockStorageState := NewMockStorageState(ctrl)
+	mockStorageState.EXPECT().Lock()
+	mockStorageState.EXPECT().Unlock()
+	mockStorageState.EXPECT().TrieState(&trie.EmptyHash).Return(mockTrieState, nil)
+
+	mockBlockImportHandler := NewMockBlockImportHandler(ctrl)
+	mockBlockImportHandler.EXPECT().HandleBlockImport(mockBlock, mockTrieState).Return(nil)
+
+	mockTelemetry := NewMockClient(ctrl)
+	mockTelemetry.EXPECT().SendMessage(gomock.Any())
+
+	s := &chainProcessor{
+		blockState:         mockBlockState,
+		storageState:       mockStorageState,
+		blockImportHandler: mockBlockImportHandler,
+		telemetry:          mockTelemetry,
+	}
+	err = s.handleBlock(block)
+	assert.NoError(t, err)
+}
+
+func Test_chainProcessor_handleBlock_nilBlock(t *testing.T) {
+	t.Parallel()
+
+	s := &chainProcessor{}
+	err := s.handleBlock(nil)
+	assert.EqualError(t, err, errors.New("block or body is nil").Error())
+}
+
+func Test_chainProcessor_handleBody(t *testing.T) {
+	t.Parallel()
+
+	var testExtrinsic = []types.Extrinsic{{1, 2, 3}, {7, 8, 9, 0}, {0xa, 0xb}}
+
+	ctrl := gomock.NewController(t)
+
+	mockTransactionState := NewMockTransactionState(ctrl)
+	mockTransactionState.EXPECT().RemoveExtrinsic(testExtrinsic[0])
+	mockTransactionState.EXPECT().RemoveExtrinsic(testExtrinsic[1])
+	mockTransactionState.EXPECT().RemoveExtrinsic(testExtrinsic[2])
+
+	testBody := types.NewBody(testExtrinsic)
+
+	type fields struct {
+		transactionState TransactionState
+	}
+	type args struct {
+		body *types.Body
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+	}{
+		{
+			name: "base case",
+			fields: fields{
+				transactionState: mockTransactionState,
```

Comment:
nit you could have instead a slice of `removedExtrinsics` and then configure the mock transaction state in the subtest, to avoid using the parent test `t` for the mock controller.

Reply:
Moved mock to subtest.